client/client.go

```diff
@@ -0,0 +1,322 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+// DefaultLogger is a default func to print logs
+func DefaultLogger(format string, args ...interface{}) {
+	log.Printf("[DEBUG] "+format, args...)
+}
+
+// RoundTripper satisfies the http.RoundTripper interface and is used to
+// customize the default http client RoundTripper
+type RoundTripper struct {
+	// Default http.RoundTripper
+	Rt http.RoundTripper
+	// Additional request headers to be set (not appended) in all client
+	// requests
+	headers *http.Header
+	// A pointer to a map of headers to be masked in logger
+	maskHeaders *map[string]struct{}
+	// A custom function to format and mask JSON requests and responses
+	FormatJSON func([]byte) (string, error)
+	// How many times HTTP connection should be retried until giving up
+	MaxRetries int
+	// If Logger is not nil, then RoundTrip method will debug the JSON
+	// requests and responses
+	Logger func(string, ...interface{})
```

Comment:
Loggers should already be concurrent/thread safe:

* https://golang.org/src/log/log.go?s=2378:2670#L40
* https://github.com/sirupsen/logrus/blob/5bd5a315a50f7f5663063a3f608757fdca721f7f/logger.go#L12-L42

I think having the previous implementation with a printf interface is sufficient.

Reply:
Yes, but I wanted to cover cases, when Logger has been unset in other thread. From my perspective, using interface was too much. Brought it back.

Reply:
Today it appeared that interface doesn't allow to specify custom inline logger func.

E.g. I want to customize logger for multiple clients:

```
                    func client(name string) client {
                        ...
                        provider.HTTPClient = http.Client{
                                Transport: &client.RoundTripper{
                                        Rt:     &http.Transport{},
                                        Logger: func(format string, args ...interface{}) {
                                            log.Debugf("[%s] "+format, name, args...)
                                        },
                                },
                        }
```

I didn't find a way to do this with interfaces.

@jtopjian did I miss something?

Reply:
Correct, custom inline functions are not interfaces. You need to create a struct that implements a `Printf` method. The struct can be as customized as you like, even have no fields, have lots of different methods, just as long as it has a `Printf` method. The `Printf` method can leverage other loggers, possibly combine logging from multiple clients, etc.

An alternative we can do is:

```go
type RoundTripper struct {
  ...
  Logger *io.Writer
}
```

Which would provide the same kind of functionality, but on a lower-level. For example, the noop logger could be removed and instead of [this](https://github.com/gophercloud/utils/blob/master/client/client.go#L163), we could do:

```
log := ioutil.Discard
if rt.Logger != nil {
  log = *rt.Logger
}

log.Printf("OpenStack Request URL: %s %s", request.Method, request.URL)
```

Reply:
Sorry for not being clear. The problem that I need to create a dynamic (inline) interface with a custom method. It looks that it is impossible with interfaces, therefore I'd like to bring back a function.

Reply:
see https://stackoverflow.com/questions/23134439/go-anonymous-method-for-a-structure

Reply:
I'd like to hear more details about why this can't be solved using standard interfaces and methods. 

Using the sample code you gave above, wouldn't this work?

```go

type MyLogger struct {
  Client string
}

func (mylogger) Printf(format, args... interface{}) {
  log.Printf("[%s] "+format, mylogger.Client, args...)
}

func client(name string) client {
  myLogger := MyLogger{
    Client: name,
  }

  provider.HTTPClient = http.Client{
    Transport: &client.RoundTripper{
      Rt:     &http.Transport{},
      Logger: myLogger,
    },
  }
}
```

I think the area I'm failing to understand is a situation where the input to the logger is going to be so dynamic that it can't be modeled in a struct. In theory, you could even do this:

```go
struct MyLogger struct {
  Foo interface{}
  Bar func()
}

func (ml) Printf(...) {
  // Treat Foo as a slice
  // or a string
  // or something else
  //
  // or call ml.Bar()
}
```

Reply:
Thanks for your hints.I figured out how to achieve what I want:

```go
type myLogger struct {
        Prefix string
}

func (l myLogger) Printf(format string, args ...interface{}) {
        log.Printf("%s [DEBUG] "+format, append([]interface{}{l.Prefix}, args...)...)
}

func clientFunc(name string) clientResult { 
    ...
    provider.HTTPClient = http.Client{
            Transport: &client.RoundTripper{
                    Rt:     &http.Transport{},
                    Logger: &myLogger{Prefix: name},
            },
    }
    ...
}
```